bench/BenchAC.hs

```diff
@@ -32,115 +40,171 @@ evalBS (ByteString xs) = eval xs M.empty
 
-
-hashCircuit
-    :: forall n p
-    .  PrimeField (Zp p)
-    => SHA2 "SHA256" (ArithmeticCircuit (Zp p)) n
-    => IO (ByteString 256 (ArithmeticCircuit (Zp p)))
-hashCircuit = do
-    x <- randomIO
-    let acX = fromConstant (x :: Integer) :: ByteString n (ArithmeticCircuit (Zp p))
-        h = sha2 @"SHA256" @(ArithmeticCircuit (Zp p)) acX
-
-    evaluate . force $ h
+hashCircuit :: forall n p .
+   PrimeField (Zp p) =>
+   SHA2 "SHA256" (ArithmeticCircuit (Zp p)) n =>
+   IO (ByteString 256 (ArithmeticCircuit (Zp p)))
+hashCircuit = sha2 @"SHA256" @(ArithmeticCircuit (Zp p)) @n . fromConstant <$> randomIO @Integer
 
 -- | Generate random addition circuit of given size
---
-additionCircuit :: forall n p. (KnownNat n, PrimeField (Zp p)) => IO (ByteString n (ArithmeticCircuit (Zp p)))
-additionCircuit = do
-    x <- randomIO
-    y <- randomIO
-    let acX = fromConstant (x :: Integer) :: ByteString n (ArithmeticCircuit (Zp p))
-        acY = fromConstant (y :: Integer) :: ByteString n (ArithmeticCircuit (Zp p))
-        acZ = from (from acX + from acY :: UInt n (ArithmeticCircuit (Zp p)))
-
-    evaluate . force $ acZ
-
-benchOps :: forall n p. (KnownNat n, PrimeField (Zp p)) => Benchmark
-benchOps = env (additionCircuit @n @p) $ \ac ->
-    bench ("Adding ByteStrings of size " <> show (value @n) <> " via UInt") $ nf evalBS ac
-
-benchHash
-    :: forall n p
-    .  PrimeField (Zp p)
-    => SHA2 "SHA256" (ArithmeticCircuit (Zp p)) n
-    => Benchmark
-benchHash = env (hashCircuit @n @p) $ \ac ->
-    bench ("Calculating SHA2 512/364 of a bytestring of length " <> show (value @n)) $ nf evalBS ac
+additionCircuit :: forall n p .
+    (KnownNat n, PrimeField (Zp p)) =>
+    IO (ByteString n (ArithmeticCircuit (Zp p)))
+additionCircuit = from @(UInt n (ArithmeticCircuit (Zp p))) <$$> (+)
+    <$> do from @(ByteString n (ArithmeticCircuit (Zp p))) . fromConstant <$> randomIO @Integer
+    <*> do from @(ByteString n (ArithmeticCircuit (Zp p))) . fromConstant <$> randomIO @Integer
+
+leqCircuit :: Ord (Bool c) (FieldElement c) => FieldElement c -> FieldElement c -> Bool c
+leqCircuit x y = x <= y
+
+fibonacciIndexCircuit :: forall c . (Ring (FieldElement c), Eq (Bool c) (FieldElement c), Conditional (Bool c) (FieldElement c)) => Integer -> FieldElement c -> FieldElement c
+fibonacciIndexCircuit nMax x = foldl (\m k -> bool m (fromConstant @Integer @(FieldElement c) k) (fib k one one == x :: Bool c)) zero [1..nMax]
+    where
+        fib :: Integer -> FieldElement c -> FieldElement c -> FieldElement c
+        fib 1 x1 _  = x1
+        fib n x1 x2 = fib (n - 1) x2 (x1 + x2)
+
+reverseListCircuit :: forall t n . Vector n t -> Vector n t
+reverseListCircuit = unsafeToVector . reverse . fromVector
+
+benchCircuit :: forall n p .
+    KnownNat n =>
+    String ->
+    IO (ByteString n (ArithmeticCircuit (Zp p))) ->
+    Benchmark
+benchCircuit desc crct = env (crct >>= evaluate . force) $ bench title . nf evalBS where
```

Comment:
What does this function benchmark?

Reply:
@TurtlePU Whole circuit evaluation. I just generalized some functions for benchmarking that looked almost the same.